clippy_lints/src/unnecessary_reserve.rs

```diff
@@ -0,0 +1,121 @@
+}
+
+impl_lint_pass!(UnnecessaryReserve => [UNNECESSARY_RESERVE]);
+
+pub struct UnnecessaryReserve {
+    msrv: Msrv,
+}
+impl UnnecessaryReserve {
+    pub fn new(msrv: Msrv) -> Self {
+        Self { msrv }
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessaryReserve {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'tcx>) {
+        if !self.msrv.meets(msrvs::CHECK_UNNECESSARY_RESERVE) {
+            return;
+        }
+
+        if let ExprKind::MethodCall(PathSegment { ident: method, .. }, struct_calling_on, args_a, _) = expr.kind
+                && method.name.as_str() == "reserve"
+                && acceptable_type(cx, struct_calling_on)
+                && let Some(block) = get_enclosing_block(cx, expr.hir_id)
+                && let Some(next_stmt_span) = check_extend_method(cx, block, struct_calling_on, &args_a[0])
+                && !next_stmt_span.from_expansion()
+            {
+                span_lint_and_then(
+                    cx,
+                    UNNECESSARY_RESERVE,
+                    next_stmt_span,
+                    "unnecessary call to `reserve`",
+                    |diag| {
+                        diag.span_suggestion(
+                            expr.span,
+                            "remove this line",
+                            String::new(),
+                            Applicability::MaybeIncorrect,
+                        );
+                    }
+                );
+            }
+    }
+
+    extract_msrv_attr!(LateContext);
+}
+
+#[must_use]
+fn acceptable_type(cx: &LateContext<'_>, struct_calling_on: &rustc_hir::Expr<'_>) -> bool {
+    let acceptable_types = [sym::Vec, sym::VecDeque];
+    acceptable_types.iter().any(|&acceptable_ty| {
+        match cx.typeck_results().expr_ty(struct_calling_on).peel_refs().kind() {
+            ty::Adt(def, _) => cx.tcx.is_diagnostic_item(acceptable_ty, def.did()),
+            _ => false,
+        }
+    })
+}
+
+#[must_use]
+fn check_extend_method(
+    cx: &LateContext<'_>,
+    block: &Block<'_>,
+    struct_expr: &rustc_hir::Expr<'_>,
+    args_a: &rustc_hir::Expr<'_>,
+) -> Option<rustc_span::Span> {
+    let args_a_kind = &args_a.kind;
+    let mut read_found = false;
+    let mut spanless_eq = SpanlessEq::new(cx);
+
+    let _: Option<!> = for_each_expr(block, |expr| {
+        if let ExprKind::MethodCall(_, struct_calling_on, _,_) = expr.kind
+            && let Some(expr_def_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id)
+            && let ExprKind::MethodCall(PathSegment { ident: method_call_a, .. },..) = args_a_kind
+            && method_call_a.name == rustc_span::sym::len
+            && match_def_path(cx, expr_def_id, &paths::ITER_EXTEND)
+            && acceptable_type(cx, struct_calling_on)
+            // Check that both expr are equal
+            && spanless_eq.eq_expr(struct_calling_on, struct_expr)
+        {
+            read_found = true;
+        }
+        let _ = !read_found;
```

Comment:
What does this do?

Reply:
Oh, trying to flip `read_found`.  I will refactor this